File: FinRegistry/logistic_regression.py
# !conda activate jupyter_env
# !cd familial_analysis
# !python3
import tqdm
import json
import numpy as np
import pandas as pd
from statsmodels.discrete import conditional_models
from FinRegistry.basic_tools import who_dict
from plot_tools import plot_odds_ratio, plot_crossed_odds_ratio, draw_grouped_bar_plot

OUTCOME = 'M13_RHEUMA'
data = pd.read_csv('data_'+OUTCOME+'.csv')
eps = json.load(open('eps_'+OUTCOME+'.json', 'r'))


def c_model(dataset, ep_index, who):
    ep_col_name = who_dict[who]+'_ep'+str(ep_index)
    try:
        endpoint = eps[ep_index]
        n_cases = dataset[ep_col_name].sum()
        lr = conditional_models.ConditionalLogit(endog=dataset.outcome,
                                                 exog=dataset[ep_col_name],
                                                 groups=dataset.subclass).fit(disp=0)
        pval = lr.pvalues[0]
        se = lr.bse[0]
        hr_025 = np.exp(lr.conf_int().iloc[0,0])
        hr_975 = np.exp(lr.conf_int().iloc[0,1])
        return [endpoint, who, se, pval, hr_025, hr_975, n_cases]
    except KeyError:
        return []


def model_loop(dataset, note, res_df):
    for i in tqdm.tqdm(range(len(eps))):
        res_fa = c_model(dataset, i, 'father')
        res_mo = c_model(dataset, i, 'mother')
        if res_fa:
            res_df = res_df.append(pd.Series(res_fa + [note], index=res.columns), ignore_index=True)
        if res_mo:
            res_df = res_df.append(pd.Series(res_mo + [note], index=res.columns), ignore_index=True)
    return res_df


res = pd.DataFrame(columns=["endpoint", "who", "se", "pval", "hr_025", "hr_975", "n_cases", "note"])
res = model_loop(data, 'all', res)
res = model_loop(data[data.sex == 1], 'women', res)
res = model_loop(data[data.sex == 0], 'men', res)
# summary statistics after removing those whose parents were born after 1961
data_sub = data[(data.fa_year_range < 1960) & (data.mo_year_range < 1960)]
res = model_loop(data_sub, 'all_sub', res)
res = model_loop(data_sub[data_sub.sex == 1], 'women_sub', res)
res = model_loop(data_sub[data_sub.sex == 0], 'men_sub', res)
# save the results
res.to_csv('res_'+OUTCOME+'.csv', index=None)

# plot the results
plot_odds_ratio(res[res.note == 'all'], eps, OUTCOME, bar_delta=.05)
plot_odds_ratio(res[res.note == 'women'], eps, OUTCOME, bar_delta=.05)
plot_odds_ratio(res[res.note == 'men'], eps, OUTCOME, bar_delta=.05)
plot_crossed_odds_ratio(res, ('women', 'men'), OUTCOME)

plot_odds_ratio(res[res.note == 'all_sub'], eps, OUTCOME, bar_cap=.05)
plot_crossed_odds_ratio(res, ('all', 'all_sub'), OUTCOME)

draw_grouped_bar_plot(res[res.note == 'all'], 'who', 'n_cases', title='All individuals')
draw_grouped_bar_plot(res[res.note == 'all_sub'], 'who', 'n_cases', title='Individuals whose parents were born < 1962')

File: FinRegistry/logistic_regression.R
setwd("~/familial_analysis")
library(dplyr)
library(survival)
library(rjson)

OUTCOME = "T1D_STRICT"
m.data <- read.csv(paste0(OUTCOME,"/data_",OUTCOME,".csv"))
m.data$subclass <- as.factor(m.data$subclass)
eps <- fromJSON(file=paste0(OUTCOME,"/eps_",OUTCOME,".json"))


get_stats <- function(who, number , note, dataframe) {
    ep_col_name <- paste0(substring(who, 1, 2),'_ep',as.character(number-1))
    data <- select(dataframe, ep_col_name, "ch_year", "mo_year", "fa_year", "number_of_sib", "outcome", "subclass")
    names(data)[names(data) == ep_col_name] <- "exposure"
    n_cases <- sum(data$exposure)
    if (n_cases < 20) {
        res <- c(eps[number],who,NaN,NaN,NaN,NaN,NaN,note,n_cases)
    } else {
        model <- clogit(outcome ~ exposure + ch_year + mo_year + fa_year + number_of_sib + strata(subclass), data, method = "exact")
        se <- summary(model)$coeff["exposure","se(coef)"]
        pval <- summary(model)$coeff["exposure","Pr(>|z|)"]
        hr <- summary(model)$conf.int["exposure","exp(coef)"]
        hr_025 <- summary(model)$conf.int["exposure","lower .95"]
        hr_975 <- summary(model)$conf.int["exposure","upper .95"]
        res <- c(eps[number],who,se,pval,hr,hr_025,hr_975,note,n_cases)
    }
    return(res)
}


results <- NULL

note = "all"
for (number in seq(1,length(eps))){
    # loop for the diseases from mothers
    results <- rbind(results,get_stats("mother", number , note, m.data))
    # loop for the diseases from fathers
    results <- rbind(results,get_stats("father", number , note, m.data))
}

note = "boy"
for (number in seq(1,length(eps))){
    # loop for the diseases from mothers
    results <- rbind(results,get_stats("mother", number , note, m.data %>% filter(sex==0)))
    # loop for the diseases from fathers
    results <- rbind(results,get_stats("father", number , note, m.data %>% filter(sex==0)))
}

note = "girl"
for (number in seq(1,length(eps))){
    # loop for the diseases from mothers
    results <- rbind(results,get_stats("mother", number , note, m.data %>% filter(sex==1)))
    # loop for the diseases from fathers
    results <- rbind(results,get_stats("father", number , note, m.data %>% filter(sex==1)))
}

RES <- data.frame(results)
colnames(RES) <- c("endpoint","who","se","pval","hr","hr_025","hr_975","note","n_cases")
write.csv(RES,file=paste0(OUTCOME, "/results_",OUTCOME,"_r.csv"), row.names = FALSE)
